### ISSUE TITLE
"generic only used once in function signature" false positive (and wrong error message) because usages in the function body should count as well

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#1540]
### Description

From @decorator-factory:

Here's a (synthetic) example of where it arguably makes a difference: 
```py
class Banana:
    def clone(self) -> Self:
        return self

class Clonable(Protocol):
    def clone(self) -> Self: ...

def clone_n_times_and_send[C: Clonable](
    obj: C,
    n: int,
    net: NetworkThing,
    banana: Banana,
) -> None:
    _ = banana.clone()
    clones: list[C] = []
    for _ in range(n):
        clones.append(obj.clone())
    net.send(clones)
```
If I said `clones: list[Clonable] = []`, I could then mistakenly do `clones.append(banana.clone())`
